DESIGN.md

```diff
@@ -0,0 +1,471 @@
+- `devinfo`: A helper to provide device info. Like what /dev to set the IO cgroup limits on. This will be determined by grabbing the device where the root (/) dir is mounted.  (ie. /dev/sda might be 8:0)
+- `cgroups`: Actually creates and deletes cgroups v2 limits (implementation described below)
+- `resources`: A helper package to abstract away all the many low level resource requirements (TBD if needed)
+
+#### Cgroups
+
+##### Notes
+- Cgroups v2 will be used. Kernel version 5.2+ is required for this functionality
+- Upon job termination the Job Controller via `joblib` will cleanup the cgroup files
+
+#### Details
+1. Create a custom cgroup directory `/sys/fs/cgroup/jobs/<jobid>`
+  - Create `/sys/fs/cgroup/jobs` if it doesnt exist yet
+2. Inside  `<jobid>` directory:
+  - set `memory.max` in bytes (one token)
+  - set `cpu.max` to "<quota> <period>" where they are in microseconds. (2 tokens)
+    - cpu.max is written as "<quota> <period>" in microseconds. Typical period is 100000µs (100ms), and quota ≤ period.
+  - set `io.max` to "<device> <rbps> <wbps>" (3 tokens). An example: `8:0 rbps=1048576 wbps=1048576`
+    - Note: The major(8) and minor(0) will try to be determined based on device where the root(/) directory is mounted on.
+3. Immediately after `ForkExec` returns (in the parent), write the child PID to `/sys/fs/cgroup/jobs/<jobid>/cgroup.procs`.
+4. When the process exits, delete the cgroup directory `/sys/fs/cgroup/jobs/<jobid>`.
+5. If the process is still running, write `1` to `/sys/fs/cgroup/jobs/<jobid>/cgroup.kill` to forcefully kill the process and remove the cgroup.
+
+
+### 3. `server` (gRPC API)
+
+- Implements:
+  - `StartJob` (takes command, resource limits)
+  - `StopJob`
+  - `GetStatus`
+  - `StreamOutput` (streaming gRPC)
+- Secures:
+  - mTLS setup using TLS 1.3
+  - authorization done via CA. O=admin|user
+
+### 4. `CLI` (JobCtl)
+
+- Command line interface using [`spf13/cobra`](https://github.com/spf13/cobra)
+- gRPC TLS client with `x509` cert loading
+- Usage is listed above
+- On `StreamOutput`, the CLI will read and when the end of the stream is reached, it will stop reading and exit. If the job is still running, it will continue to read in real-time until the job exits or the user interrupts the stream.
+
+
+### 5. AuthN/Z
+- The client must authenticate the server **and** the server must authenticate the client. This means that the client **and** the server must trust each other via CA's. For simplicity, the server will have the client's CA's when running.
+- `CN` will be used for `<username>`
+- `O` will be used for `<role>-<group>`
+- Process streaming or getting statuses, reading, will be limited to members of any said group. This will be for isolation purposes.
+- Depending on the `role` (admin|user) , the running of particular commands will be limited.
+
+#### NOTE: All lists will be hardcoded in the implementation
+
+### Denylist for `admin` users and normal users
```

Comment:
I'd consider alternate approaches. Allow/deny lists are really easy to circumvent - there's just too many corner cases.

For example: `ln /bin/bash /home/me/myunrestrictedshell` or `apt install some-shell-thats-not-on-thelist` and then the deny list is completely ineffective from there.